libmscore/easeInOut.cpp

```diff
@@ -0,0 +1,116 @@
+    qreal d = x + 3.0 * w1 - 3.0 * w2 + (1.0 - x);
+    qreal a = (-3.0 * x - 6.0 * w1 + 3 * w2) / d;
+    qreal b = (3.0 * x + 3.0 * w1) / d;
+    c = -x / d;
+    qreal a2 = a * a;
+    qreal p = (3.0 * b - a2) / 3.0;
+    qreal q = (2.0 * a2 * a - 9.0 * a * b + 27.0 * c) / 27.0;
+    qreal discr = (q * q) / 4.0 + (p * p * p) / 27.0;
+    qreal t = 0.0;
+    if (discr < 0.0) {
+        // Case that happens about 75% of the time --> 3 roots.
+        qreal mp3 = -p / 3.0;
+        qreal r = std::sqrt(mp3 * mp3 * mp3);
+        qreal phi = std::acos(std::min(std::max(-1.0, -q / (2.0 * r)), 1.0));
+        qreal t1 = 2.0 * std::pow(r, 1.0 / 3.0);
+        // Find the only root that is within the 0 to 1 interval.
+        t = t1 * std::cos(phi / 3.0) - a / 3.0;
+        if (0.0 > t || t > 1.0) {
+            t = t1 * std::cos((phi + 2.0 * pi) / 3.0) - a / 3.0;
+            if (0.0 > t || t > 1.0) {
+                t = t1 * std::cos((phi + 4.0 * pi) / 3.0) - a / 3.0;
+            }
+        }
+    } else if (discr > 0.0) {
+        // Case that happens about 25% of the time --> 1 root.
+        qreal q2 = q / 2.0;
+        qreal sd = std::sqrt(discr);
+        qreal u = std::pow(-q2 + sd, 1.0 / 3.0);
+        qreal v = std::pow(q2 + sd, 1.0 / 3.0);
+        t = u - v - a / 3.0;
+    } else {
+        // Case that happens extremely rarely --> 2 roots.
+        qreal q2 = q / 2.0;
+        qreal u = q2 < 0.0 ? std::pow(-q2, 1.0 / 3.0) : -std::pow(q2, 1.0 / 3.0);
+        // Find the only root that is within the 0 to 1 interval.
+        t = 2.0 * u - a / 3.0;
+        if (0.0 > t || t > 1.0) {
+            t = -u - a / 3.0;
+        }
+    }
+    return t;
+}
+
+//-------------------------------------------------------------------------------------------------
+// This function is even more simplified and optimized because all the Bezier control points are
+// constant. Thus, when computing the Y root there is only one case and the math simplifies to the
+// following simple expression.
+//-------------------------------------------------------------------------------------------------
+qreal EaseInOut::tFromY(const qreal y) const
+{
+      if (y == 0.0 || y == 1.0)
+            return y;
+      return 0.5 + std::cos((4.0 * M_PI + std::acos(1.0 - 2.0 * y)) / 3.0);
+      }
+
+//-------------------------------------------------------------------------------------------------
+// Given a number of note to place within the given duration, return the list of on-times for each
+// note given the current ease-in and ease-out parameters. The first note is at time 0 while the 
+//-------------------------------------------------------------------------------------------------
+void EaseInOut::timeList(const int nbNotes, const int duration, QList<int>* times) const
+{
+      qreal nNotes = qreal(nbNotes);
+      qreal space = qreal(duration);
+      if (_easeIn == 0.0 && _easeOut == 0.0) {
+            for (qreal n = 0; n <= nNotes; n++) {
```

Comment:
Better to keep the loop variable integer.

Reply:
I actually made a change to 'for (qreal n = 0.0; n <= nNotes; n += 1.0) {' to make the intention more specific.

Reply:
You might be lucky here but if the step is 1.3 and <code>nNotes</code> is 7.8 you might have one iteration too little because if cumulative errors in the add operations. The common way to handle is to have an integer loop variable and calculate the real value inside the loop. E.g.
```
for (int i = 0; i < nNotes; ++i) {
     real n = static_cast<qreal>(i);
     ...
}
````

Also, it might be my Fortran background (oops!), but a variable <code>n</code> is for me an integer, not a float :wink: .

Reply:
Well. OK. No more arguments from me. I will just do that.